app/src/main/java/org/tokend/template/features/signin/UnlockAppActivity.kt

```diff
@@ -0,0 +1,301 @@
+        canSignIn = password_edit_text.text.isNotEmpty()
+                && !password_edit_text.hasError()
+                && email.isNotEmpty()
+    }
+
+    private fun updateSignInButtonState() {
+        unlock_button.isEnabled = canSignIn
+    }
+
+    private fun updateFieldsLoadingState(isLoading: Boolean) {
+        val animatedView =
+                if (usePassword) password_layout
+                else fingerprint_layout
+
+        if (isLoading) {
+            AnimationUtil.fadeOutView(sign_out_button)
+            AnimationUtil.fadeOutView(
+                    view = animatedView,
+                    viewGone = false)
+            progress.show()
+        } else {
+            AnimationUtil.fadeInView(animatedView)
+            AnimationUtil.fadeInView(sign_out_button)
+            progress.hide()
+        }
+    }
+
+    private fun requestFingerprintAuthIfAvailable() {
+        fingerprintAuthManager.requestAuthIfAvailable(
+                onAuthStart = { usePassword = false },
+                onSuccess = { email, password ->
+                    signIn(email, password)
+                },
+                onError = {
+                    toastManager.short(it)
+                }
+        )
+    }
+
+    private fun cancelFingerprintAuth() {
+        fingerprintAuthManager.cancelAuth()
+    }
+
+    private fun tryToSignIn() {
+        when {
+            password_edit_text.text.isEmpty() ->
+                password_edit_text.setErrorAndFocus(R.string.error_cannot_be_empty)
+            canSignIn -> {
+                SoftInputUtil.hideSoftInput(this)
+                signIn(email, password_edit_text.text.getChars())
+            }
+        }
+    }
+
+    private fun signIn(email: String, password: CharArray) {
+        SignInUseCase(
+                email,
+                password,
+                apiProvider.getKeyServer(),
+                session,
+                credentialsPersistor,
+                PostSignInManager(repositoryProvider)
+        )
+                .perform()
+                .compose(ObservableTransformers.defaultSchedulersCompletable())
+                .doOnSubscribe {
+                    loadingIndicator.setLoading(true)
+                    SoftInputUtil.hideSoftInput(this)
+                }
+                .doOnEvent {
+                    password.fill('0')
+                }
+                .doOnDispose { password.fill('0') }
+                .subscribeBy(
+                        onComplete = {
+                            Navigator.from(this).toMainActivity()
+                        },
+                        onError = {
+                            it.printStackTrace()
+                            handleUnlockError(it)
+                            loadingIndicator.setLoading(false)
+                        }
+                )
+                .addTo(compositeDisposable)
+    }
+
+    private fun handleUnlockError(error: Throwable) {
+        when (error) {
```

Comment:
If error occurred during fingerprint unlock (connection error for example) we must reenable fingerprint auth